ADBDEV 1061

tsdkv opened the pull request:
`gplogfilter` outputs multiline text, which is ambiguously perceived by csv parsers.

`gplogfilter`'s result line:
```
2020-08-29 00:06:46.635873 UTC|data_mart|reporting|p158978|th1580669056|10.0.39.10|33084|2020-08-29 00:01:37 UTC|0|con750679|cmd6|seg8|slice1||||LOG: |00000|duration: 308909.520 ms||||||
SET optimizer=OFF;
INSERT INTO data_mart.volatility_15min(time_interval, active, instrument_type, volatility_date, volatility, volume, pnl)
WITH ti AS (
  SELECT generate_series(
            data_mart.get_last_volatility_date() - interval '1 hour',
            date_trunc('hour',now()) - INTERVAL '15 minutes',
            INTERVAL '15 minutes'
        ) time_interval
)
SELECT
    t.time_interval,
    t.active,
    d.instrument_type,
    date,
    volatility,
    sum(volume) / 1000000.0 volume,
    sum(pnl) / 1000000.0    pnl
FROM (
    SELECT
        time_interval,
        active,
        (VALUES->'900'->>'value')::numeric volatility,
        instant date,
        row_number() OVER (PARTITION BY active, time_interval ORDER BY instant DESC) r
       FROM volatility.history h
           JOIN ti ON
        h.instant > ti.time_interval and h.instant <= ti.time_interval + '15 minutes'
       ) t
    JOIN (
        SELECT
            ti.time_interval,
            instrument_active_id active,
            instrument|0||postgres.c|1421|
```

maksm90 commented:
I think it's wrong to make modification of some field values explicitly removing newline characters. @t1mursadykov try to preserve these values adding quotes around multiline strings

maksm90 commented on gpMgmt/bin/gplogfilter:
```diff
@@ -460,8 +460,8 @@ try:
-            # Write filtered lines to output file.  Don't append \n to
-            # each line, because the original line ends are still there.
+            # Write filtered lines to output file. There is a multi-line
+            # text containing \n, and lot of unnecessary consecutive spaces.
```
Is this comment line still actual?

maksm90 commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -273,6 +273,11 @@ def next(self):
         # our single regex will match both.
         item[16] = item[16] + ": "
+
+        for i in range(len(item)):
```
You have to prepend this block by comment that you quotes **multiline** string

maksm90 commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -273,6 +273,11 @@ def next(self):
         item[16] = item[16] + ": "
+
+        for i in range(len(item)):
+            if item[i].count('\n'):
```
Use more readable predicate here: `'\n' in item[i]`

maksm90 commented:
I have encountered another issue with `gplogfilter` utility: when query contains csv delimiter character `|`, e.g. the following:
```sql
select a | 1 from test;
```
then csv generated by `gplogfilter` is mistaken that this delimiter separates columns of csv.
@t1mursadykov try to fix also this issue.

maksm90 commented:
another issue if query text contains double-quote, e.g., around irregular table's object name:
```sql
select * from "test";
```
then in csv query preserves these quotes but after importing it would miss their. I suppose we have to double quote these kind of queries.

maksm90 commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -273,6 +273,13 @@ def next(self):
+        # There is a text containing \n or | symbols. We need to add quotation 
+        # marks, for parsers to process csv text correctly.
+        for i in range(len(item)):
+            if '\n' in item[i] or '|' in item[i]:
```
@t1mursadykov pls check that `|` is the only possible csv delimiter for gplogfilter`s output

maksm90 replied:
you might use another more readable predicate for verifying whether some character is in string - by [link](https://stackoverflow.com/questions/19463556/string-contains-any-character-in-group#answer-19463605 )

tsdkv replied:
Do we really need to use another method to check for a character in the string? I think the current way is readable.

maksm90 replied:
I think we might assign `|` symbol to some variable to explicitly name it and to prepare to easy delimiter modification further.

t1mursadykov commented:
@maksm90, `|` is the only possible csv delimiter

t1mursadykov commented:
Something like csv standard
https://www.rfc-editor.org/rfc/rfc4180.txt

maksm90 commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -273,6 +273,17 @@ def next(self):
+
+        # There is a text containing \n or | symbols. We need to add quotation 
+        # marks, for parsers to process csv text correctly. We also need to 
+        # double quote in places where they already are.
```
1. There are trailing spaces in lines of comment. Take care of it. You might setup necessary plugin to your editor or check it via `git diff` before commiting patch 
2. Rewrite comment that we have to quote such column values that contain new line, csv delimiter or double quotation marks. The last ones have to be escaped by another double quote. All this have to be done to generate valid csv.

maksm90 commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -273,6 +273,17 @@ def next(self):
+        for i in range(len(item)):
+            if '\n' in item[i] or '|' in item[i]:
+                if '\"' in item[i]:
+                    item[i] = item[i].replace('\"', '\"\"')
```
If item contains double-quote we have to quote overall string anyway. You could check the result of exporting csv line without enclosing quotes - postgres removes internal quotes

maksm90 reviewed (approved):
Nonetheless, take a look at using `any()` expression instead of multiple `or`s in predicate of symbols containment. And remove new line before quotation of `item[i]` so this quotation builds a single block up with previous loop and its segregation is redundant.

darthunix commented:
Hmmm.. And have you tried triple quotes in python?

darthunix commented:
@t1mursadykov can you show some example of the problem you are fixing in this PR? I have checked `gplogfilter` on MacOS with several query examples and didn't succeed to get fixed lines as in an example of @maksm90:
```sql
select
'a',
'|',
'"',
'b'
from generate_series(1,2);

select 'a' || '|c' || '"b';
```
May be it would be a good idea to add some tests (behave tests?)

leskin-in commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -47,6 +47,8 @@
 import sys
 import time
 
+csv_delimeter = '|'
```
Use lowerCamelCase for variable names. This is de-facto codestyle of this file.

t1mursadykov commented:
We cannot use `csv.writer` because this object is not a generator. We can't get ready csv strings from it.

darthunix commented:
> We cannot use `csv.writer` because this object is not a generator. We can't get ready csv strings from it.

@t1mursadykov well, allow me to retort:
```python
import csv
import cStringIO

def iterable_csv():
    with open('1.txt', 'rb') as file:
        reader = csv.reader(file)
        buffer = cStringIO.StringIO()
        writer = csv.writer(buffer, quoting=csv.QUOTE_ALL)
        for row in reader:
            writer.writerow(row)
            line = buffer.getvalue()
            yield line
            buffer.truncate(0)

iter = iterable_csv()
for line in iter:
    print(line)
```
I suppose you can get rid of a custom CSV parser and modify the [yeilded](https://github.com/arenadata/gpdb/blob/a5b7bb85595badf7f2dd11e717206f75b7d3214e/gpMgmt/bin/gppylib/logfilter.py#L205) lines in `FilterLogEntries()` with a `cStringIO` buffer from the example above.

Commit:
Fix gplogfilter csv generation

The result of gplogfilter is ambiguously perceived
by parsers. To fix this, the standard csv.writer
class is used to generate csv.

t1mursadykov commented:
Changed the `CsvFlatten` class as showed above by @darthunix. The standard `csv.writer` class now is  used to generate the json.

darthunix commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -42,11 +42,15 @@
 import sys
 import time
 
+csvDelimeter = '|' 
```
A trailing whitespace in the end of the line.

darthunix commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -261,19 +265,25 @@ class CsvFlatten(object):
         self.source = iter(iterable)
-    
+        self.buffer = cStringIO.StringIO()
+        self.writer = csv.writer(self.biffer, delimiter=csvDelimeter, quotechar='"', quoting=csv.QUOTE_MINIMAL)
```
Typo in `self.biffer` (expecting `self.buffer` instead)

darthunix commented on gpMgmt/bin/gppylib/logfilter.py:
```diff
@@ -261,19 +265,25 @@ class CsvFlatten(object):
-        return '|'.join(item) + "\n"
+
+        self.buffer.truncate(0)
+        self.writer.write(item)
```
`csv.writer` doesn't have a  metod `write()` (use `writerow()` instead).

darthunix reviewed (changes requested):
At the moment PR doesn't work (look at review comments). But I have applied fixes and tested logfilter for the queries:
```sql
select
'a',
'|',
'"',
'b'
from generate_series(1,2);

select 'a' || '|c' || '"b';

select 1 | 1 from generate_series(1, 2);
```
Output:
```
2021-02-03 13:45:36.803160 +10|darthunix|adb|p94301|th328904192|127.0.0.1|55165|2021-02-03 13:45:35 +10|0|con13|cmd1|seg-1||||sx1|LOG: |00000|"statement: select
'a',
'|',
'""',
'b'
from generate_series(1,2);"||||||"select
'a',
'|',
'""',
'b'
from generate_series(1,2);"|0||postgres.c|1639|
2021-02-03 13:45:41.955846 +10|darthunix|adb|p94301|th328904192|127.0.0.1|55165|2021-02-03 13:45:35 +10|0|con13|cmd3|seg-1||||sx1|LOG: |00000|"statement: select 'a' || '|c' || '""b';"||||||"select 'a' || '|c' || '""b';"|0||postgres.c|1639|
2021-02-03 13:45:42.221767 +10|darthunix|adb|p94301|th328904192|127.0.0.1|55165|2021-02-03 13:45:35 +10|0|con13|cmd5|seg-1||||sx1|LOG: |00000|"statement: select 1 | 1 from generate_series(1, 2);"||||||"select 1 | 1 from generate_series(1, 2);"|0||postgres.c|1639|
```
It seems to be correct. But the errors should be fixed to approve this PR.

maksm90 commented:
@t1mursadykov @darthunix workflow to verify result csv from `gplogfilter` utility:

1. execute testing query
2. save last logs into temp csv file:
  ```bash
  gplogfilter -n NUM_LAST_ROWS -q /path/to/active/log_file.csv > /tmp/log.csv
  ```
3. try to import result csv to test table:
  ```sql
  create temp table test as select * from gp_toolkit.gp_log_system limit 0;
  \copy test from /tmp/log.csv with (format csv, delimiter '|');

  # check fields with SQL statements, they have to correspond to input
  select logmessage, logdebug from test;
  ```
